### DISCUSSION TITLE
ERR_INVALID_URL when deploying Next.js project that use Hygraph Content API stored in environment variable

### DISCUSSION THREAD

**Discussion by naufalHaidar12342** [vercel/next.js#69165] [Help]
### Link to the code that reproduces this issue

https://github.com/naufalHaidar12342/hygraph-nextjs14-test

### To Reproduce

1. Push project to GitHub using Git
2. Connect GitHub repo to Vercel
3. Add environment variable in Vercel that holds Hygraph Content API endpoint
4. Add environment variable in Vercel that holds Hygraph Permanent Authentication Token
5. Build/Deploy project in Vercel
6. An `ERR_INVALID_URL` will be thrown by Vercel if the Hygraph Content API only stored in environment variable

### Current vs. Expected behavior

Current behavior:
- Next.js project will failed to build and throws `ERR_INVALID_URL` if the Hygraph Content API only stored in environment variable
- Next.js project could be built successfully IF there is a fallback API. This fallback API is just a variable in javascript, exported, that contains the identical Hygraph Content API
- Hygraph Permanent Authentication Token somehow able to be parsed by Vercel during deployment

Expected behavior:
- Next.js project should be able to read environment variable that contains/hold Hygraph Content API (Hygraph High Performance Content API)

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 32672
  Available CPU cores: 8
Binaries:
  Node: 20.16.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: 14.2.5
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Upstream

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

### Additional context

This issues has been happening since Next.js 13 App Router. I only use `stable` version of Next.js for all of my projects. After discussion with both Vercel team and Hygraph team, one of Vercel staff suggest me to open this issues. You can find my discussion on Vercel Community here:

[ERR_INVALID_URL on Hygraph Public API stored in environment variable](https://vercel.community/t/err-invalid-url-on-hygraph-public-api-stored-in-environment-variable/213)

[ERR_INVALID_URL on Hygraph Content API](https://vercel.community/t/err-invalid-url-on-hygraph-content-api/427)

For my discussion with Hygraph team, I even conducted a simple test where I will create new nextjs project (nextjs 14, accesible through the repo link) and one of Hygraph team member will do the same. We share the API and the auth token to verify that this issue does persist in other user, specifically when the next.js project being deployed to Vercel. If the same project being run locally, there is no issue with the environment variable

**Comment by samcx** [comment]
@naufalHaidar12342

Have you tried console logging any Environment Variable from :vercel: and checking your logs after (ideally in a Route Handler, that's not static)?

**Reply by naufalHaidar12342** [reply]
I apologize before since I still don't understand how to do it. Can you give me some references or examples on how to console log the Environment Variables (i.e. articles, discussion)? Thank you 🙏

**Reply by samcx** [reply]
Just anywhere in your React component!

```
console.log(`env variable = ${process.env.HYGRAPH_CONTENT_API}`).
```

**Reply by naufalHaidar12342** [reply]
Ahh thanks for clarifying it, I thought it has something to do with Vercel CLI 😂
After adding console.log to log the environment variable that stores the Hygraph Content API, the environment variable shows up in Vercel deployment log (in development branch)

**Reply by samcx** [reply]
So the Environment Variable is present, but you are still receiving that error `ERR_INVALID_URL`?

**Reply by naufalHaidar12342** [reply]
sadly yes 😭. I will try calling the environment variable again in one fetch function, then push it to production branch and see if its still causing error

**Reply by naufalHaidar12342** [reply]
Hi, just want to check in and says the error still persist. I tested the environment variable in certain async fetch function, push it to my development branch, and it failed to build on Vercel. Here is a snippet from the build log of my project. I edit the Hygraph URL to prevent it being abused and as a proof of error

```
TypeError: Failed to parse URL from "https://ap-southeast-2.cdn.hygraph.com/content/xxxxxxxxxxxx/master"
    at new Request (node:internal/deps/undici/undici:6110:19)
    at i (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:177224)
    ... 5 lines matching cause stack trace ...
    at stringify (<anonymous>)
    at eR (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:134889)
    at eP (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:135332) {
  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL
      at new NodeError (node:internal/errors:405:5)
      at new URL (node:internal/url:676:13)
      at new Request (node:internal/deps/undici/undici:6108:25)
      at i (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:177224)
      at b (/vercel/path0/.next/server/app/page.js:1:9855)
      at P (/vercel/path0/.next/server/app/page.js:45:128)
      at em (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:131226)
      at /vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:142926
      at Array.toJSON (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:146504)
      at stringify (<anonymous>) {
    input: '"https://ap-southeast-2.cdn.hygraph.com/content/xxxxxxxxxxxxxx/master"',
    code: 'ERR_INVALID_URL'
  }
```

**Reply by samcx** [reply]
@naufalHaidar12342 Not entirely sure what's going on here.

Is there a Hygraph guide/template to try testing this (where do I get my own Environment Variable)?

**Reply by naufalHaidar12342** [reply]
So far, I have not found any Hygraph blog post that shows how to replicate this issue. But, if you want to test it yourself, I created a simple nextjs proejct that fetch data using async function just on the homepage (`page.jsx`). But you have to create your own environment variable in Vercel and in `.env.local`
[nextjs14+hygraph simple test project](https://github.com/naufalHaidar12342/hygraph-nextjs14-test)

Just remove the `Authorization` props inside the header, since it is optional

**Reply by samcx** [reply]
@naufalHaidar12342 Unfortunately this isn't probably going to work because I still need a `process.env.HYGRAPH_CONTENT_API` to test.

Can you create a simpler :repro: without any data logic, so I can test with my own HYGRAPH_CONTENT_API env I can obtain from Hygraph's dashboard?

**Reply by naufalHaidar12342** [reply]
sorry for the late reply. just to make sure, that means i will create nextjs project from `npx create-next-app` and create `.env` file with variable named `HYGRAPH_CONTENT_API` ? thank you!

**Reply by samcx** [reply]
@naufalHaidar12342 Yes!

**Reply by naufalHaidar12342** [reply]
I hope this is the right repo for your testing [Next.js App Router with env variable placeholder](https://github.com/naufalHaidar12342/nextjs-hygraphenv-test)

**Reply by samcx** [reply]
![CleanShot 2024-09-04 at 09 53 50@2x](https://github.com/user-attachments/assets/8e831662-8743-4045-9f1c-bc0ecc7a2fcc)

@naufalHaidar12342 So the env is coming back fine for me—it's unclear to me why your URL is considered invalid.

**Reply by naufalHaidar12342** [reply]
I assume that env log generated from your production branch in Vercel? Does that mean something wrong with my Vercel account or is this could be an edge case?

**Reply by samcx** [reply]
@naufalHaidar12342 In my example, I'm not using the URL in a fetch, so something could be off with how you're trying to use it. This just confirms there isn't something wrong with Vercel injecting the Environment Variable.
